Week3/homework/6-step3.js

```diff
@@ -1,6 +1,18 @@
 'use strict';
 
 const arr2d = [[1, 2], [3, 4], [5, 6]];
-const arr3d = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]];
 
-// add your solution here, or add a comment on how you would tackle this problem
+console.log(arr2d.join('\n') + '\n\n');
+
+function arr3d(arr2d) {
+  const rowCount = arr2d.length;
+  const rowSizes = []
+
+  for (let i = 0; i < rowCount; i++) {
+    rowSizes.push(arr2d[i].length)
+  }
+  return [rowCount, Math.min.apply(null, rowSizes)]
+}
+
+arr3d([[[1, 2], [3, 4]], [[5, 6], [7, 8]]]);
+console.log('The length of array is : ' + (arr2d.length));
```

Comment:
I think you might have missed the point of this exercise.
The idea is to go into each of the arrays and its "sub" arrays find the values (i.e. not arrays) and do something with them (i.e. print the value out).

So, your function is printing each row, but not finding the rows within the rows and printing the values from them.  The best way to tackle this is using a technique called `recursion`.  What that does, is test for some condition, and if it is true, the function recalls itself with a subset of the original value. It is best seen in code:
```
const arr2d = [[1, 2], [3, 4], [5, 6]];
const arr3d = [[[1, 2], [3, 4]], [[5, 6], [7, 8]]];
const arrNd = [[[[1, 2]], [3, 4], [[[5, 6], [7, 8]], [9, 10]], [11, 12]], [[[[[[13, 14]]]]]]];

function printAllTheItems(arr){ 
  let values = [];
  for (let i = 0; i < arr.length; i++){
    if (Array.isArray(arr[i])) {
      values = values.concat(printAllTheItems(arr[i]));
    } else {
      values = values.concat([arr[i]]);
    }
  }
  return values;
};

console.log(printAllTheItems(arr2d));
console.log(printAllTheItems(arr3d));
console.log(printAllTheItems(arrNd));
```
Actually this flattens the array (using for loops) and then prints the results of flattening each of the arr2d, arr3d, and arr4d.
As I've seen from your last question you are not afraid of reduce, so here is a reduce based version:
```
const flatten = (arr, result=[]) => {
  return arr.reduce((acc, elem) => {
    if (Array.isArray(elem)) {
      return acc.concat(flatten(elem, result));
    } else {
      return acc.concat([elem]);
    }
  }, result);
}

flatten(arrNd);
```
The important point in both of these is that there is a test to see whether the current item is an Array.  If it is, we know we have to go into the array, so we call the function again, and pass the current item (an array) into the function.  We then `concat` the return of that call onto our `acc` (accumulator or results array), and keep going.

The nice thing with recursion, and what makes it so important, is that it won't care how deep the arrays are nested in arrays, it will just keeps going collecting values.

I hope this gave some insights. 😄